bokehjs/src/lib/models/glyphs/webgl/markers.ts

```diff
@@ -107,42 +72,58 @@ export class MarkerGL extends BaseGLGlyph {
     if (this._show == null)
       this._show = new Uint8Buffer(this.regl_wrapper)
 
-    const prev_nmarkers = this._show.length
-    const show_array = this._show.get_sized_array(nmarkers)
-    if (indices.length < nmarkers) {
-      this._show_all = false
-
-      // Reset all show values to zero.
-      for (let i = 0; i < nmarkers; i++)
-        show_array[i] = 0
-
-      // Set show values of markers to render to 255.
-      for (let j = 0; j < indices.length; j++) {
-        show_array[indices[j]] = 255
+    const ntypes = mainGlGlyph._unique_marker_types.length
+    for (let i = 0; i < ntypes; i++) {
+      const marker_type = mainGlGlyph._unique_marker_types[i]
+      if (marker_type == null)
+        continue
+
+      let nshow = nmarkers  // Number of markers to show.
+      const prev_nmarkers = this._show.length
+      const show_array = this._show.get_sized_array(nmarkers)
+      if (ntypes > 1 || indices.length < nmarkers) {
+        this._show_all = false
+
+        // Reset all show values to zero.
+        for (let i = 0; i < nmarkers; i++)
+          show_array[i] = 0
+
+        // Set show values of markers to render to 255.
+        nshow = 0
+        for (let j = 0; j < indices.length; j++) {
+          const k = indices[j]  // Marker index.
```

Comment:
Unless you need `j` for other purpose, it's simpler and faster to use this:
```suggestion
        for (const k of indices) {
```